### PR TITLE
Add Identity Update Builder

### PR DESCRIPTION
**Pull request by neekolas**
## tl;dr

- Adds `IdentityUpdateBuilder` and `SignatureRequest` structs, used to create batched Identity Updates

## Design decisions

There were a few constraints on the design that made this a tricky one to implement.

- These are batch signatures. An IdentityUpdate may contain several actions, each requiring one or two signatures. But each of those signatures are against a signature text derived from the entire batch. So, we need to gather an outline of all the updates before we can start collecting signatures.
- Passing signers over the FFI barrier has proven complicated in the past. Some signers are synchronous, some are asynchronous. The caller doesn't necessarily have all the signers available every time they use `libxmtp`, so they may have to prompt the user before they can get a signature. The most versatile solution is to have a "signature request" that goes over the FFI barrier to platform SDKs, and have the SDKs return the serialized bytes.

## Usage

```rust
        let account_address = "0x1234".to_string();
        let nonce = 0;
        let inbox_id = generate_inbox_id(&account_address, &nonce);

        let mut signature_request = IdentityUpdateBuilder::new(inbox_id)
            .create_inbox(account_address.into(), nonce)
            .add_association("0x5678".to_string().into(), account_address.into())
            .to_signature_request();

        for missing_signature in signature_request.missing_signatures() {
            signature_request.add_signature(somehow_get_signature_maybe_this_happens_over_ffi()).expect("should succeed")?;
        }

        let identity_update = signature_request.build_identity_update()?;
```

### CODE REVIEW

**Comment by neekolas** [comment]
### Merge activity

* **Apr 8, 12:10 PM EDT**: [Graphite](https://app.graphite.dev/github/pr/xmtp/libxmtp/632) rebased this pull request after merging its parent, because this pull request is set to merge when ready.

**Review by richardhuaaa — APPROVED** [review]
This is a really elegant way to implement batch signatures. Seconding @insipx's comments to make it easier for others to understand.

Also, from the example code:
```
        for missing_signature in signature_request.missing_signatures() {
            signature_request.add_signature(somehow_get_signature_maybe_this_happens_over_ffi()).expect("should succeed")?;
        }
```

I think we can build higher-level API's over this before exposing to FFI. Rough sketch:

```
register_inbox_and_installation_key(Option<RecoveryAddress>): TextToSign; // installation key signature handled internally. Does not register inbox if inbox already exists
associate_external_wallet(): TextToSign; // installation key signature handled internally
... etc.
```

**Comment by richardhuaaa** [comment]
Oof, didn't realize it was auto-merge @neekolas

File: xmtp_id/src/associations/builder.rs [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+    pending_signatures: HashMap<SignatureField, MemberIdentifier>,
+}
+
+pub struct IdentityUpdateBuilder {
+    inbox_id: String,
+    client_timestamp_ns: u64,
+    actions: Vec<PendingIdentityAction>,
+}
+
+impl IdentityUpdateBuilder {
```
[the remark this answers is not in the document]
Comment (neekolas) [reply xmtp_id/src/associations/builder.rs:42]: I think that's a good suggestion. I'll rename.

File: xmtp_id/src/associations/builder.rs [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+};
+
+#[derive(Error, Debug)]
+pub enum IdentityBuilderError {
+    #[error("Missing signer")]
+    MissingSigner,
+}
+
+#[derive(Clone, PartialEq, Hash, Eq)]
+enum SignatureField {
```
[the remark this answers is not in the document]
Comment (neekolas) [reply xmtp_id/src/associations/builder.rs:23]: Agree it's not the most intuitive. Batch signatures in general are a little confusing.

At the very least, I'll add some clarifying comments. And if we find a better way to structure the code we can refactor later.
